packages/elements-core/src/utils/ref-resolving/resolvedObject.ts

```diff
@@ -0,0 +1,73 @@
+import { isArray, isPlainObject } from 'lodash';
+
+import { defaultResolver, ReferenceResolver } from './ReferenceResolver';
+
+const resolvedObjectSymbol = Symbol('ResolvedObject');
+const originalObjectSymbol = Symbol('OriginalObject');
+
+interface CreateResolvedObjectOptions {
+  contextObject?: object;
+  resolver?: ReferenceResolver;
+}
+
+export const createResolvedObject = (currentObject: object, options: CreateResolvedObjectOptions = {}): object =>
+  recursivelyCreateResolvedObject(currentObject, currentObject, [], options);
+
+const recursivelyCreateResolvedObject = (
+  currentObject: object,
+  rootCurrentObject: object,
+  propertyPath: string[],
+  options: CreateResolvedObjectOptions = {},
+): object => {
+  const mergedOptions = {
+    contextObject: options.contextObject || currentObject,
+    resolver: options.resolver || defaultResolver(options.contextObject || currentObject),
+  };
+  if (!currentObject || currentObject[resolvedObjectSymbol]) return currentObject;
+
+  const cachedValues = {};
+
+  return new Proxy(currentObject, {
+    get(target, name) {
+      if (name === resolvedObjectSymbol) return true;
+
+      if (name === originalObjectSymbol) return currentObject;
+
+      if (cachedValues[name]) return cachedValues[name];
+
+      const value = target[name];
+
+      const newPropertyPath = [...propertyPath, name.toString()];
+
+      let resolvedValue;
+      if (isPlainObject(value) && value.hasOwnProperty('$ref')) {
+        try {
+          resolvedValue = mergedOptions.resolver(
+            { pointer: value['$ref'], source: null },
```

Comment:
There's a tiny tiny risk that `value['$ref']` (which can be `value.$ref` bc `value is any`) is not a `string`. That can crash a perfectly written resolver. You could maybe replace the `value.hasOwnProperty('$ref')` with `typeof value.$ref === 'string'` or similar.